File: services/server/src/protocol/communication/packet.py
from protocol.messages import message_codes, bet, ask_winners, batch
from protocol.common import errors

PACKET_MIN_LEN = 2
PACKET_LENGTH_BYTES = 2

class Packet:
    def __init__(self, message_code, message):
        self.message_code = message_code
        self.message = message

    def to_bytes(self):
        buf = bytearray()
        buf.append(self.message_code)
        buf.extend(self.message.to_bytes())
        return bytes(buf)


    @classmethod
    def from_bytes(cls, data):
        if len(data) < PACKET_MIN_LEN:
            raise ValueError(errors.PACKET_TOO_SHORT_ERR)
        offset = 0
        message_code = int(data[offset])
        offset += 1

        try:
            match message_code:
                case message_codes.BATCH_CODE:
                    message = batch.Batch.from_bytes(data[offset:])
                case message_codes.ASK_WINNERS_CODE:
                    message = ask_winners.AskWinners.from_bytes(data[offset:])
                case _:
                    raise ValueError(errors.UNEXPECTED_MESSAGE_ERR)

        except Exception as e:
            raise ValueError(f"{errors.DESERIALIZE_PACKET_ERR}: {e}")

        return cls(message_code, message)

File: services/client/src/protocol/communication/packet.go
package communication

import (
	"fmt"

	errors "github.com/7574-sistemas-distribuidos/tp-nivelador/src/protocol/common"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/protocol/messages"
)

const PACKET_MIN_LEN = 1

type Packet struct {
	MessageCode uint8
	Message     messages.Message
}

func NewPacket(messageCode uint8, message messages.Message) Packet {
	return Packet{
		MessageCode: messageCode,
		Message:     message,
	}
}

func (p *Packet) ToBytes() []byte {
	messageBytes := p.Message.ToBytes()
	messageLen := uint16(len(messageBytes))

	totalLen := PACKET_MIN_LEN + messageLen
	bytes := make([]byte, 0, totalLen)

	bytes = append(bytes, byte(p.MessageCode))
	bytes = append(bytes, messageBytes...)

	return bytes
}

func PacketFromBytes(data []byte) (Packet, error) {
	if len(data) < PACKET_MIN_LEN {
		return Packet{}, fmt.Errorf(errors.PacketTooShortError)
	}
	offset := 0

	messageCode := int(data[offset])
	offset++

	var message messages.Message
	var err error
	switch messageCode {
	case messages.BET_CODE:
		message, err = messages.BetFromBytes(data[offset:])
	case messages.ASK_WINNERS_CODE:
		message, err = messages.AskWinnersFromBytes(data[offset:])
	case messages.WINNER_CODE:
		message, err = messages.BetFromBytes(data[offset:])
	case messages.FINISH_CODE:
		message, err = messages.FinishFromBytes(data[offset:])
	default:
		return Packet{}, fmt.Errorf(errors.UnexpectedMessage)
	}
	if err != nil {
		return Packet{}, err
	}

	return Packet{MessageCode: uint8(messageCode), Message: message}, nil
}
